src/Sema.zig

```diff
@@ -5512,16 +5512,156 @@ fn zirArrayCat(sema: *Sema, block: *Scope.Block, inst: Zir.Inst.Index) CompileEr
+                const val = try Value.Tag.array.create(anon_decl.arena(), buf);
+                return sema.analyzeDeclRef(try anon_decl.finish(
+                    ty,
+                    val,
+                ));
+            }
+            return sema.mod.fail(&block.base, lhs_src, "TODO array_cat more types of Values", .{});
+        } else {
+            return sema.mod.fail(&block.base, lhs_src, "TODO runtime array_cat", .{});
+        }
+    } else {
+        return sema.mod.fail(&block.base, lhs_src, "TODO runtime array_cat", .{});
+    }
+}
+
+const ArrayCatInfo = struct { array_elem_type: Type, sent: ?Value = null, len: u64 };
+fn getArrayCatInfo(t: Type) ?ArrayCatInfo {
+    var info: ArrayCatInfo = .{ .array_elem_type = undefined, .len = undefined };
+    if (t.zigTypeTag() == .Array) {
+        info.len = t.arrayLen();
+        if (t.tag() == .array_u8) {
+            info.array_elem_type = Type.initTag(.u8);
+        } else if (t.tag() == .array_u8_sentinel_0) {
+            info.array_elem_type = Type.initTag(.u8);
+            info.sent = Value.initTag(.zero);
+        } else if (t.tag() == .array) {
+            const payload = t.castTag(.array).?.data;
+            info.array_elem_type = payload.elem_type;
+        } else if (t.tag() == .array_sentinel) {
+            const payload = t.castTag(.array_sentinel).?.data;
+            info.array_elem_type = payload.elem_type;
+            info.sent = payload.sentinel;
+        }
+    } else if (t.zigTypeTag() == .Pointer) {
+        const ptrinfo = t.ptrInfo().data;
+        if (ptrinfo.pointee_type.zigTypeTag() != .Array) return null;
+        if (ptrinfo.size != .One) return null;
+        return getArrayCatInfo(ptrinfo.pointee_type);
+    } else return null;
+    return info;
 }
 
 fn zirArrayMul(sema: *Sema, block: *Scope.Block, inst: Zir.Inst.Index) CompileError!Air.Inst.Ref {
     const tracy = trace(@src());
     defer tracy.end();
 
-    _ = inst;
-    return sema.mod.fail(&block.base, sema.src, "TODO implement zirArrayMul", .{});
+    const inst_data = sema.code.instructions.items(.data)[inst].pl_node;
+    const extra = sema.code.extraData(Zir.Inst.Bin, inst_data.payload_index).data;
+    const lhs = sema.resolveInst(extra.lhs);
+    const rhs = sema.resolveInst(extra.rhs);
+    const lhs_ty = sema.typeOf(lhs);
+    const lhs_src: LazySrcLoc = .{ .node_offset_bin_lhs = inst_data.src_node };
+    const rhs_src: LazySrcLoc = .{ .node_offset_bin_rhs = inst_data.src_node };
+
+    // In `**` rhs has to be comptime-known, but lhs can be runtime-known
+    const tomulby = try sema.resolveInt(block, rhs_src, extra.rhs, Type.initTag(.usize));
+    const mulinfo = getArrayCatInfo(lhs_ty) orelse
+        return sema.mod.fail(&block.base, lhs_src, "expected array, found '{}'", .{lhs_ty});
+
+    const final_len = std.math.mul(u64, mulinfo.len, tomulby) catch return sema.mod.fail(&block.base, rhs_src, "operation results in overflow", .{});
```

Comment:
Speaking from painful experience working on a compiler with first class multidimensional array support, this is something you should try to avoid.

This essentially just blows up your AST without any benefit. You will perform all following semantic analysis / transformations that come afterwards essentially `tomulby` times.

If there is great caching of arbitrary transformations in place great, but in general you would really want to have a special purpose ZIR / AIR node that es essentially a binop between a value and a dimension.

That considerably flattens the AST at the cost that you have a second array type that you need to consider.

Reply:
Thanks for the advice @miscco. This code has nothing to do with the AST but I think your point still stands. Even so, this expansion code is needed, since comptime code could, for example, modify the array at an arbitrary index. So what you are describing would be a performance optimization to make the compiler more lazy. That's a good idea, but it can come later.

Reply:
AST was essentially a shorthand for a tree based representation of the program (even if it is stored in flat memory)

It is essentially an optimization and fully depends on whether it becomes an actual problem. just something to keep in the back of your head when expanding a single node into many

Reply:
Yeah, I have in mind some optimizations for this, but I will put them in a separate pr and benchmark them first.